---
title: "exactTestDGE_nhi"
author: "LBaer"
date: "11/06/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## R setup

```{r}
# Load libraries
library(limma)
library(edgeR)
library(AnnotationHub)
library(tidyverse)
library(magrittr)
library(scales)
library(pander)
library(ggrepel)
library(here)
```

## Data setup

```{r}
# Load Nhi's DGEList
nhiDGE <- read_rds("nhiData/dge_g.rds") %>%
  .[,rownames(subset(.$samples, Age == 6 & Hypoxia == 0))]
```

```{r}
# Set group variable
nhiDGE$samples$group <- colnames(nhiDGE) %>%
  str_extract("(w|q)") %>%
  factor(levels = c("w", "q"))
```

## Add gene information

```{r}
# Add AnnotationHub and subset to search for zebrafish
ah <- AnnotationHub()
ah %>%
  subset(species == "Danio rerio") %>%
  subset(dataprovider == "Ensembl") %>%
  subset(rdataclass == "EnsDb")
```

```{r}
# Select correct Ensembl release
ensDb <- ah[["AH64906"]]
```

```{r}
# Extract GenomicRanges object from ensDb
genesGR <- genes(ensDb)
```

```{r}
# Remove redundant columns from mcols
mcols(genesGR) <- mcols(genesGR)[c("gene_id", "gene_name", "gene_biotype", "entrezid")]
```

```{r}
# Add genesGR to DGEList using rownames of DGEList to reorder the genesGR
nhiDGE$genes <- genesGR[rownames(nhiDGE),]
```

## Data QC

```{r}
# Perform logical test to see how many genes were not detected in dataset
nhiDGE$counts %>%
  rowSums() %>%
  is_greater_than(0) %>%
  table()
```

```{r}
# Check for genes having > 4 samples with cpm > 1
nhiDGE %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(4) %>%
  table()
```

```{r}
# Create logical vector of genes to keep that fit criteria
genes2keep <- nhiDGE %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(4)
```

```{r}
# Create new DGEList of genes fitting criteria
nhiFilt <- nhiDGE[genes2keep,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
# Clean up samples element
nhiFilt$samples <- nhiFilt$samples[,1:3]
```

```{r}
# Compare distributions of the DGELists before and after filtering
par(mfrow = c(1,2))
nhiDGE %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "Before Filtering")
nhiFilt %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "After Filtering")
par(mfrow = c(1,1))
```

## Library sizes

```{r}
# Check library sizes with box plot
nhiFilt$samples %>%
  ggplot(aes(group, lib.size, fill = group)) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  labs(x = "Genotype", y = "Library Size") +
  scale_fill_discrete(name ="Genotype", labels = c("Wildtype",            "Mutant")) +
  scale_x_discrete(labels=c("w" = "Wildtype", "q" = "Mutant")) +
    theme_bw()
```

## PCA

```{r}
# Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
pca <- nhiFilt %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```
 
```{r}
# Quick inspection to check whether first two PCA components capture most of the variability
summary(pca)$importance %>% pander(split.tables = Inf)
```

```{r}
# Plot PCA
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(nhiFilt$samples, "sample")) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point() +
  geom_text_repel() +
  theme_bw()
```

## Differential expression

```{r}
# Create model matrix
design <- model.matrix(~group, data = nhiFilt$samples)
```

```{r}
# Perform exact test on DGEList
topTable <- nhiFilt %>%
  estimateDisp(design = design) %>%
  exactTest() %>%
  topTags(n = Inf) %>%
  .$table %>%
  as_tibble() %>%
  unite("Range", ID.start, ID.end, sep = "-") %>%
  unite("Location", ID.seqnames, Range, ID.strand, sep = ":") %>%
  dplyr::select(Geneid = ID.gene_id, 
                 Symbol = ID.gene_name,
                 AveExpr = logCPM, logFC, 
                 P.Value = PValue, 
                 FDR, Location, 
                 Entrez = ID.entrezid) %>%
  mutate(DE = FDR < 0.05)
```