---
title: "Thesis export code"
author: "Lachlan Baer"
date: "09/10/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# R Setup

```{r}
# Set working directory
setwd("~/Documents/Bioinformatics/Projects/20190122_Q96K97_NoStress_RNASeq/R")
```

```{r options}
theme_set(theme_bw())
panderOptions("table.split.table", Inf)
panderOptions("big.mark", ",")
panderOptions("table.style", "rmarkdown")
if (interactive()) setwd(here::here("R"))
```

# Export code for thesis

```{r}
#ggsave
ggsave("~/Documents/Honours 2019/5. Thesis/Thesis results/pcaRUVs_2.png",
       scale = 0.75)
```

```{r}
png("~/Documents/Honours 2019/5. Thesis/Thesis results/appendices/A.4.1.png", 
    units = "cm", width = 15.44, height = 10, res = 300)
dev.off()
```

```{r}
# Function for rounding numeric columns of dataframe
roundDF <- function(df) {
  nums <- vapply(df, is.numeric, FUN.VALUE = logical(1))
  df[,nums] <- lapply(
    df[,nums], 
    formatC, 
    digits = 4, 
    format = "f", 
    drop0trailing = TRUE
  )
  (df)
}
```

```{r, eval=FALSE}
# DE 2019 Table
write.table(topTableDE %>% 
              dplyr::select(-P.Value) %>% 
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/DEAnalysis2019.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r, eval=FALSE}
# DE 2017 Table
write.table(topTable2017DE %>% 
              dplyr::select(-t, -p, -DE) %>% 
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/appendices/C.1.1.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
# GSEA Hallmark 2019 Table
write.table(fgseaHallmarkTop %>% 
              dplyr::select(-pval, -FDR, -nMoreExtreme, -leadingEdge) %>%
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/fgseaHallmark2019.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
# GSEA KEGG 2019 Table
write.table(fgseaKEGGTop %>% 
              dplyr::select(-pval, -FDR, -nMoreExtreme, -leadingEdge) %>%
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/appendices/B.1.1.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
# GSEA Hallmark 2019 RUVs Table
write.table(fgseaHallmarkRUVs2019Top %>% 
              dplyr::select(-pval, -FDR, -nMoreExtreme, -leadingEdge) %>%
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/fgseaHallmarkRUVs2019.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
# GSEA Hallmark Integrated RUVs Table
write.table(fgseaHallmarkRUVsTop %>% 
              dplyr::select(-pval, -FDR, -nMoreExtreme, -leadingEdge) %>%
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/fgseaHallmarkRUVsComb.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
#GSEA Hallmark 2017 Table
write.table(fgseaHallmarkTop2017 %>% 
              dplyr::select(-pval, -FDR, -nMoreExtreme, -leadingEdge) %>%
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/fgseaHallmark2017.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r, eval=FALSE}
# DE Table of integrated data after RUVSeq
write.table(topTableRUVsDE %>% 
              dplyr::select(-P.Value) %>% 
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/appendices/E.3.1.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r, eval=FALSE}
# DE 2019 Table after RUVSeq
write.table(topTableRUVs2019DE %>% 
              dplyr::select(-P.Value) %>% 
              roundDF(), 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/appendices/E.4.1.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
# metaData
write.table(metaData, 
            file = "~/Documents/Honours 2019/5. Thesis/Thesis results/appendices/D.1.1.txt",
            quote = FALSE,
            sep = ",",
            row.names = FALSE)
```

```{r}
# PCA comparisons 
pcaComb$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(dgeFiltComb$samples, "sample")) %>%
  mutate(
    data = case_when(
      str_detect(.$sample, "_") == TRUE ~ "2017",
      str_detect(.$sample, "_") == FALSE ~ "2019"
    )
  ) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample, shape = data)) +
  geom_point(alpha = 0.8, size = 3) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaComb)$importance[2,"PC1"]), ")"),
    y =  paste0("PC2 (", percent(summary(pcaComb)$importance[2,"PC2"]), ")"),
    colour = "Genotype",
    shape = "Dataset"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant")) +
  scale_shape_manual(values = c(17, 16)) +
  coord_cartesian(ylim = c(-20, 35), xlim = c(-80, 65)) +
  theme_bw()
pcaRemoved$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(dgeRemoved$samples, "sample")) %>%
  mutate(
    data = case_when(
      str_detect(.$sample, "_") == TRUE ~ "2017",
      str_detect(.$sample, "_") == FALSE ~ "2019"
    )
  ) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample, shape = data)) +
  geom_point(alpha = 0.8, size = 3) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaRemoved)$importance[2,"PC1"]), ")"),
    y =  paste0("PC2 (", percent(summary(pcaRemoved)$importance[2,"PC2"]), ")"),
    colour = "Genotype",
    shape = "Dataset"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant")) +
  coord_cartesian(ylim = c(-20, 35), xlim = c(-80, 65)) +
  theme_bw()
```

```{r}
# PCA comparisons 2019
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(dgeFilt$samples, "sample")) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  # geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(pca)$importance[2, "PC1"]), ")"),
    y = paste0("PC2 (", percent(summary(pca)$importance[2, "PC2"]), ")"),
    colour = "Genotype"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant")) +
  coord_cartesian(ylim = c(-40, 40), xlim = c(-50, 20))
pcaRemoved2019$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(dgeRemoved2019$samples, "sample")) %>%
  mutate(
    data = case_when(
      str_detect(.$sample, "_") == TRUE ~ "2017",
      str_detect(.$sample, "_") == FALSE ~ "2019"
    )
  ) %>%
  ggplot(aes(PC1, -PC2, colour = group, label = sample)) +
  geom_point(alpha = 0.8, size = 3) +
  # geom_text_repel(show.legend = FALSE) +
  labs(
    x = paste0("PC1 (", percent(summary(pcaRemoved2019)$importance[2,"PC1"]), ")"),
    y =  paste0("PC2 (", percent(summary(pcaRemoved2019)$importance[2,"PC2"]), ")"),
    colour = "Genotype"
  ) +
  scale_colour_discrete(labels = c("Wildtype", "Mutant")) +
  coord_cartesian(ylim = c(-40, 40), xlim = c(-50, 20))
```

# Poster render

```{r fig.align = "center", eval = FALSE}
# Create tidygraph object
tidyHallmark <- 
  tbl_graph(nodes = nodesHallmark, edges = edgesHallmark, directed = FALSE) %>%
  activate(nodes) %>%
  left_join(idConvertSymbol, by = "label") %>%
  mutate(
    pathways = case_when(
      id <= nrow(fgseaHallmarkTop) ~ label %>%
        str_replace_all("_", " ")
    ),
    DE = case_when(
      label %in% topTableDE$Geneid ~ symbol
    ),
    size = case_when(
      label %in% topTable$Geneid ~ 
        as.integer(row_number(label %in% topTable$Geneid)), 
      id <= nrow(fgseaHallmarkTop) ~ as.integer(4000)
    ),
    colour = case_when(
      id <= nrow(fgseaHallmarkTop) ~ rainbow(nrow(fgseaHallmarkTop))[id],
      label %in% topTableDE$Geneid ~ "black"
    ),
    hjust = case_when(
      DE == "ugt5b4" ~ as.integer(0)
    ),
    vjust = case_when(
      DE == "ugt5b4" ~ as.integer(5)
    )
  ) %>%
  activate(edges) %>%
  mutate(
    colour = case_when(
      from <= nrow(fgseaHallmarkTop) ~ rainbow(nrow(fgseaHallmarkTop))[from]
    )
  )
# Set seed to enable reproducibility (seed selected to create graph with non-overlapping labels)
set.seed(22)
```

```{r, eval = FALSE}
pdf("hallmarkNetwork.pdf", width = 13, height = 9.25)
```

```{r, eval = FALSE}
# Plot graph
ggraph(tidyHallmark, layout = "fr") +
  scale_fill_manual(
    values = c(rainbow(nrow(fgseaHallmarkTop)), "black"), 
    na.value = "gray80"
  ) +
  geom_edge_arc(
    aes(color = colour), 
    alpha = 0.5, 
    show.legend = FALSE, 
    curvature = 0.5
  ) +
  geom_node_point(
    aes(size = size, fill = colour),
    shape = 21,
    stroke = 0.5, 
    show.legend = FALSE
  ) +
  geom_node_label(
    aes(label = pathways),
    repel = TRUE,
    size = 3, 
    alpha = 0.7,
    label.padding = 0.15
  ) +
  geom_node_text(
    aes(label = DE, hjust = hjust, vjust = vjust),
    repel = TRUE,
    size = 3,
    alpha = 0.8,
    colour = "black"
  ) +
  theme_graph() +
  theme(legend.position = "none")
```

```{r, eval = FALSE}
dev.off()
```
