---
title: "filteredStarDGE"
author: "L Baer"
date: "23/05/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## R Setup
```{r}
library(limma)
library(edgeR)
library(AnnotationHub)
library(tidyverse)
library(magrittr)
library(scales)
library(pander)
library(ggrepel)
library(here)
```

## Data Setup
```{r}
#Load counts
counts <- read_tsv(here("4_filteredData/featureCounts/genes.out")) %>%
  set_colnames(basename(colnames(.))) %>%
  set_colnames(str_remove(colnames(.), "Aligned.sortedByCoord.out.bam"))  %>%
  dplyr::select(Geneid, starts_with("W"), starts_with("Q"))
```

```{r}
#Create DGE list and calculate normalisaton factors
dgeList <- counts %>%
  as.data.frame() %>%
  column_to_rownames("Geneid") %>%
  DGEList() %>%
  calcNormFactors()
```

```{r}
#Set group variable
dgeList$samples$group <- colnames(dgeList) %>%
  str_extract("(W|Q)") %>%
  factor(levels = c("W", "Q"))
```

```{r}
#Add gene information
ah <- AnnotationHub()
```

```{r}
#Subset Annotation Hub to search for zebrafish
ah %>%
  subset(species == "Danio rerio") %>%
  subset(dataprovider == "Ensembl") %>%
  subset(rdataclass == "EnsDb")
```

```{r}
#Select correct Ensembl release
ensDb <- ah[["AH64906"]]
```

```{r}
#Extract GenomicRanges object from ensDb
genesGR <- genes(ensDb)
```

```{r}
#Remove redundant columns from mcols
mcols(genesGR) <- mcols(genesGR)[c("gene_id", "gene_name", "gene_biotype", "entrezid")]
```


```{r}
#Use rownames of dgeList to reorder the genes object
dgeList$genes <- genesGR[rownames(dgeList),]
```

```{r}
#Perform logical test to see how many genes were not detected in dataset
dgeList$counts %>%
  rowSums() %>%
  is_greater_than(0) %>%
  table()
```
This many genes were not detected in the data set.

```{r}
#Check for genes having > 4 samples with cpm > 1
dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(4) %>%
  table()
```
This many genes have at least 4 samples with cpm > 1.

```{r}
#Create logical vector of genes to keep for further analysis
genes2keep <- dgeList %>%
  cpm() %>%
  is_greater_than(1) %>%
  rowSums() %>%
  is_weakly_greater_than(4)
```

```{r}
#Create new DGE list of filtered genes
dgeFilt <- dgeList[genes2keep,, keep.lib.sizes = FALSE] %>%
  calcNormFactors()
```

```{r}
#Compare distributions of the two datasets
par(mfrow = c(1,2))
dgeList %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "Before Filtering")
dgeFilt %>%
  cpm(log = TRUE) %>%
  plotDensities(legend = FALSE, main = "After Filtering")
par(mfrow = c(1,1))
```
Distributions of genes before and after filtering.

```{r}
#Create box plot to check library sizes
dgeFilt$samples %>%
  ggplot(aes(group, lib.size, fill = group)) +
  geom_boxplot() +
  scale_y_continuous(labels = comma) +
  labs(x = "Genotype", y = "Library Size") +
  scale_fill_discrete(name ="Genotype", labels = c("Wildtype", "Mutant")) +
  scale_x_discrete(labels=c("W" = "Wildtype", "Q" = "Mutant")) +
  theme_bw()
```
Library sizes after filtering.

```{r}
#Assess cpm values to make sure PCA results are not heavily skewed by highly expressed genes
pca <- dgeFilt %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp()
```

```{r}
#Quick inspection to check whether first two PCA components capture most of the variability
summary(pca)$importance %>% pander(split.tables = Inf)
```

```{r}
design <- model.matrix(~group, data = dgeFilt$samples)
```

```{r}
#Create object for voom analysis
voomData <- voomWithQualityWeights(dgeFilt, design = design, plot = TRUE)
```

```{r}
#Plot PCA
#plotly::ggplotly(
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  as_tibble() %>%
  dplyr::select(sample, PC1, PC2) %>%
  left_join(rownames_to_column(voomData$targets, "sample")) %>%
  ggplot(aes(PC1, PC2, colour = group, label = sample)) +
  geom_point() +
  geom_text_repel() +
  theme_bw()
#  )
```